src/mlpack/methods/ann/layer/multihead_attention.hpp

```diff
@@ -48,54 +47,51 @@ namespace mlpack {
  * of shape `(embedDim * tgtSeqLen, batchSize)`. The embeddings are stored
  * consequently.
  *
- * @tparam InputType Type of the input data (arma::colvec, arma::mat,
- *         arma::sp_mat or arma::cube).
- * @tparam OutputType Type of the output data (arma::colvec, arma::mat,
+ * srcSeqLen is inferred from dim 0 of the previous layer.
+ * embedDim is inferred from
+ *
+ * @tparam MatType Type of the input/output data (arma::colvec, arma::mat,
  *         arma::sp_mat or arma::cube).
  * @tparam RegularizerType Type of the regularizer to be used.
  */
 template <
-    typename InputType = arma::mat,
-    typename OutputType = arma::mat,
+    typename MatType = arma::mat,
     typename RegularizerType = NoRegularizer
 >
-class MultiheadAttentionType : public Layer<InputType, OutputType>
+class MultiheadAttentionType : public Layer<MatType>
 {
  public:
   /**
    * Default constructor.
    */
   MultiheadAttentionType();
 
-  // TODO: does srcSeqLen need to be given?
   /**
    * Create the MultiheadAttention object using the specified modules.
+   * The attnMask and keyPaddingMask arguments are defined as a mask
+   * where 1 passes values and 0 masks them, unlike the original
+   * paper which uses 0 to pass values and -inf to mask them.
    *
    * @param tgtSeqLen Target sequence length.
-   * @param srcSeqLen Source sequence length.
-   * @param embedDim Total dimension of the model.
    * @param numHeads Number of parallel attention heads.
-   * @param attnMask Two dimensional Attention Mask.
-   * @param keyPaddingMask Key Padding Mask.
+   * @param selfAttention Use self-attention; source key, query, and value all
+   *     come from the same inputs
    */
   MultiheadAttentionType(const size_t tgtSeqLen,
-                         const size_t srcSeqLen,
-                         const size_t embedDim,
                          const size_t numHeads,
-                         const InputType& attnmask = InputType(),
-                         const InputType& keyPaddingMask = InputType());
```

Comment:
Yeh, so I was debating on this.  My use case it actually made sense to set them later, but I guess they could still be set in the constructor.  If so, they should also be serialized, otherwise it isn't really consistent.

Reply:
Your call, like I said either is fine. Thanks for mentioning serialization, I overlooked that in the review.  Regardless of how we take them in, both members should definitely be serialized. :+1:

Reply:
I wasn't sure actually.  I'm slightly fuzzy on how people actually use/train the multiheadattention layer -- I am using it, but not sure it's quite right.  It seems to me that you actually have to modify both of these masks as you are training, so setting them in the constructor and/or serializing them doesn't exactly make sense.  But I guess if there are use-cases where the masks can just be fixed, then it would make sense to do both of those things (constructor/serialization).

Reply:
Yeah, I'm not sure that these were actually used in a subsequent `Transformer` class.  I found mlpack/models#30 (BERT) which would use this, but it depends on a `TransformerEncoder` class that I think was not finished or not merged.  My understanding of the way the attention masks are used is the same as yours, that they are set each pass.  So I'm fine with them being removed; after all it's more likely that a `Transformer` layer that handles the masking internally will be created at some point, and that wouldn't have a need to set the masks in the constructor. :+1:

Reply:
Actually, now I'm going the other way.  ;)
Can always leave them empty, but if there are cases where you want to construct a network with pre-defined masks (and save/load it), it's nice to not have to set them separately.

Reply:
Added them back into the constructor and serialization.